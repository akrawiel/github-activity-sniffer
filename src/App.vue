<template lang="pug">
#app
  .flex.font-bold.justify-center.py-4.text-2xl.text-gray-700 GitHub Activity Sniffer
  search-field(@clear="onSearchClear" @error="onSearchError" :hasError="Boolean(errorMessage)" :transformOrganizationResponse="transformOrganizationResponse")
  error-message(:message="errorMessage")
  activity-item-container
    activity-item(v-for="member in memberData" :key="member.id" :member="member")
  base-button(@click="onLoadMoreClick" v-if="nextPageLink" :disabled="isFetching") {{ isFetching ? 'Loading...' : 'Load more' }}
</template>

<script>
import ActivityItem from "@/components/ActivityItem.vue";
import ActivityItemContainer from "@/components/ActivityItemContainer.vue";
import BaseButton from "@/components/BaseButton.vue";
import ErrorMessage from "@/components/ErrorMessage.vue";
import SearchField from "@/components/SearchField.vue";
import fetchWrapper from "@/api/fetch-wrapper";

export default {
  name: "App",
  data() {
    return {
      isFetching: false,
      nextPageLink: null,
      memberData: [],
      errorMessage: ""
    };
  },
  components: {
    ActivityItem,
    ActivityItemContainer,
    BaseButton,
    ErrorMessage,
    SearchField
  },
  methods: {
    setSearchState({ nextPageLink, memberData, errorMessage }) {
      this.nextPageLink = nextPageLink;
      this.memberData = memberData;
      this.errorMessage = errorMessage;
    },
    onSearchClear() {
      this.setSearchState({
        nextPageLink: null,
        memberData: [],
        errorMessage: ""
      });
    },
    onSearchError(error) {
      this.setSearchState({
        nextPageLink: null,
        memberData: [],
        errorMessage: error
      });
    },
    async transformMembersEvents({ data }) {
      return await Promise.all(
        data.map(async member => {
          const events = await fetchWrapper({
            url: `${member.events_url.replace("{/privacy}", "")}/public`
          }).then(response => {
            if (!response.ok) {
              throw new Error(response.statusText);
            }

            return response.json();
          });

          return {
            ...member,
            latest_event: events[0] ?? null
          };
        })
      );
    },
    async transformOrganizationResponse({ promise }) {
      this.isFetching = true;
      this.errorMessage = "";

      try {
        const organizationResponse = await promise().then(async response => {
          if (!response.ok) {
            throw new Error(response.statusText);
          }

          return {
            links: response.headers.get("Link")?.split(",") ?? [],
            data: await response.json()
          };
        });

        const membersWithLatestEvents = await this.transformMembersEvents({
          data: organizationResponse.data
        });

        this.nextPageLink =
          organizationResponse.links
            ?.find(link => /rel="next"\s*$/.test(link))
            ?.match(/<([^<>]+)>/)?.[1] ?? null;

        this.memberData = [
          ...this.memberData,
          ...membersWithLatestEvents.map(
            ({ id, login, events_url, avatar_url, latest_event }) => ({
              id,
              login,
              avatar_url,
              events_url,
              latest_event
            })
          )
        ];
      } catch (error) {
        this.setSearchState({
          errorMessage: error?.message,
          nextPageLink: null,
          memberData: []
        });
      } finally {
        this.isFetching = false;
      }
    },
    async onLoadMoreClick() {
      if (!this.isFetching) {
        await this.transformOrganizationResponse({
          promise: () => fetchWrapper({ url: this.nextPageLink })
        });
      }
    }
  }
};
</script>

<style lang="postcss">
#app {
  @apply max-w-3xl p-3 flex flex-col mx-auto;
  font-family: Spartan, sans-serif;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
}

body {
  @apply bg-blue-100 min-h-screen;
}
</style>
