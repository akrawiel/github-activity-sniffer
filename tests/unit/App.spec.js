import { shallowMount } from "@vue/test-utils";
import App from "@/App.vue";

const invalidOrganizationPromise = () =>
  Promise.resolve({
    ok: false,
    statusText: "Test 1"
  });

const dummyMemberData = [
  {
    avatar_url: "avatar",
    events_url: "url",
    id: 1,
    login: "Test"
  }
];

const dummyEventData = [
  {
    id: 1
  }
];
const dummyHeaders = new Headers({
  Link: '<test-link> rel="next"'
});

const validOrganizationPromise = () =>
  Promise.resolve({
    ok: true,
    json: () => dummyMemberData,
    headers: dummyHeaders
  });

const validEventsPromise = () =>
  Promise.resolve({
    ok: true,
    json: () => dummyEventData
  });

let wrapper;

beforeEach(() => {
  wrapper = shallowMount(App);
});

describe("App.vue", async () => {
  it("renders when all data is passed", () => {
    expect(wrapper.element).toMatchSnapshot();
  });

  it("returns valid organization response data", async () => {
    await wrapper.vm.transformOrganizationResponse({
      promise: invalidOrganizationPromise
    });

    expect(wrapper.vm.errorMessage).toBe("Test 1");

    global.fetch = jest.fn(() =>
      Promise.resolve({
        ok: false,
        statusText: "Test 2"
      })
    );

    try {
      await wrapper.vm.transformMembersEvents({
        data: dummyMemberData
      });
    } catch (error) {
      expect(error).toMatchObject(new Error("Test 2"));
    }

    global.fetch = jest.fn(validEventsPromise);

    await wrapper.vm.transformOrganizationResponse({
      promise: validOrganizationPromise
    });

    expect(wrapper.vm.errorMessage).toBe("");
    expect(wrapper.vm.memberData).toMatchObject([
      {
        ...dummyMemberData[0],
        latest_event: dummyEventData[0]
      }
    ]);
    expect(wrapper.vm.nextPageLink).toBe("test-link");
  });
});
