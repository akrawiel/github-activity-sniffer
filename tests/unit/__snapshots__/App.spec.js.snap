// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App.vue renders when all data is passed 1`] = `
<div
  id="app"
>
  <div
    class="flex font-bold justify-center py-4 text-2xl text-gray-700"
  >
    GitHub Activity Sniffer
  </div>
  <search-field-stub
    transformorganizationresponse="function () { [native code] }"
  />
  <error-message-stub
    message=""
  />
  <activity-item-container-stub />
  <!---->
</div>
`;
